File: src/main/java/net/onelitefeather/pandorascluster/land/Land.java
package net.onelitefeather.pandorascluster.land;

import jakarta.persistence.*;
import net.onelitefeather.pandorascluster.enums.LandRole;
import net.onelitefeather.pandorascluster.land.player.LandMember;
import net.onelitefeather.pandorascluster.land.player.LandPlayer;
import net.onelitefeather.pandorascluster.land.position.HomePosition;
import net.onelitefeather.pandorascluster.util.ChunkUtil;
import org.bukkit.Chunk;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

@Entity
public class Land implements Cloneable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @OneToOne
    private LandPlayer owner;

    @OneToOne
    private HomePosition homePosition;

    @OneToMany(fetch = FetchType.EAGER)
    private List<LandMember> landMembers;

    @OneToMany(fetch = FetchType.LAZY, mappedBy = "land")
    private List<ChunkPlaceholder> chunks;

    @Column
    private String world;

    @Column
    private int x;

    @Column
    private int z;

    public Land() {
    }

    public Land(@NotNull LandPlayer owner, @NotNull HomePosition homePosition, @NotNull String world, int x, int z) {
        this();
        this.owner = owner;
        this.homePosition = homePosition;
        this.world = world;
        this.x = x;
        this.z = z;
        this.chunks = new ArrayList<>();
        this.landMembers = new ArrayList<>();
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    @NotNull
    public LandPlayer getOwner() {
        return owner;
    }

    public void setOwner(@NotNull LandPlayer owner) {
        this.owner = owner;
    }

    @NotNull
    public HomePosition getHomePosition() {
        return homePosition;
    }

    public void setHomePosition(@NotNull HomePosition homePosition) {
        this.homePosition = homePosition;
    }

    @NotNull
    public List<LandMember> getLandMembers() {
        return landMembers;
    }

    public void setLandMembers(@NotNull List<LandMember> landMembers) {
        this.landMembers = landMembers;
    }

    @NotNull
    public List<ChunkPlaceholder> getMergedChunks() {
        return chunks;
    }

    public void setMergedChunks(@NotNull List<ChunkPlaceholder> chunks) {
        this.chunks = chunks;
    }

    public boolean isChunkConnected(@NotNull Chunk chunk) {

        boolean connected = false;
        for (int i = 0; i < this.chunks.size() && !connected; i++) {
            ChunkPlaceholder mergedChunk = this.chunks.get(i);
            connected = ChunkUtil.getChunkIndex(chunk) == mergedChunk.getChunkIndex();
        }

        return connected;
    }

    @Nullable
    public ChunkPlaceholder getMergedChunk(long chunkIndex) {

        ChunkPlaceholder chunkPlaceholder = null;

        for (int i = 0; i < this.getMergedChunks().size() && chunkPlaceholder == null; i++) {
            ChunkPlaceholder placeholder = this.getMergedChunks().get(i);
            if (placeholder.getChunkIndex() == chunkIndex) {
                chunkPlaceholder = placeholder;
            }
        }

        return chunkPlaceholder;
    }

    @NotNull
    public String getWorld() {
        return world;
    }

    public void setWorld(@NotNull String world) {
        this.world = world;
    }

    public int getX() {
        return x;
    }

    public void setX(int x) {
        this.x = x;
    }

    public int getZ() {
        return z;
    }

    public void setZ(int z) {
        this.z = z;
    }

    @Nullable
    public LandMember getLandMember(@NotNull UUID uuid) {

        LandMember landMember = null;

        for (int i = 0; i < this.getLandMembers().size() && landMember == null; i++) {
            LandMember member = this.getLandMembers().get(i);
            if (member.getMember().getUniqueId().equals(uuid)) {
                landMember = member;
            }
        }

        return landMember;

    }

    public boolean hasAccess(@NotNull UUID uuid) {
        LandMember landMember = getLandMember(uuid);
        if (landMember == null) return false;
        return landMember.getRole().hasAccess();
    }

    public boolean isOwner(@NotNull UUID uniqueId) {
        return this.getOwner().getUniqueId().equals(uniqueId);
    }

    @Override
    public String toString() {
        return "Land{" +
                "owner=" + owner +
                ", world='" + world + '\'' +
                ", x=" + x +
                ", z=" + z +
                '}';
    }

    @Override
    @NotNull
    public Land clone() throws CloneNotSupportedException {
        return (Land) super.clone();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Land land)) return false;

        if (id != land.id) return false;
        if (x != land.x) return false;
        if (z != land.z) return false;
        if (!owner.equals(land.getOwner())) return false;
        if (!Objects.equals(homePosition, land.homePosition)) return false;
        if (!Objects.equals(landMembers, land.landMembers)) return false;
        if (!Objects.equals(chunks, land.chunks)) return false;
        return Objects.equals(world, land.world);
    }

    public boolean isMerged() {
        return !this.getMergedChunks().isEmpty();
    }

    public boolean isBanned(UUID uniqueId) {
        LandMember landMember = getLandMember(uniqueId);
        return landMember != null && landMember.getRole() == LandRole.BANNED;
    }
}

File: src/main/kotlin/net/onelitefeather/pandorascluster/land/Land.kt
package net.onelitefeather.pandorascluster.land

import jakarta.persistence.*
import net.onelitefeather.pandorascluster.enums.LandRole
import net.onelitefeather.pandorascluster.enums.Permission
import net.onelitefeather.pandorascluster.extensions.hasPermission
import net.onelitefeather.pandorascluster.land.flag.LandFlag
import net.onelitefeather.pandorascluster.land.flag.LandFlagEntity
import net.onelitefeather.pandorascluster.land.flag.getDefaultFlag
import net.onelitefeather.pandorascluster.land.player.LandMember
import net.onelitefeather.pandorascluster.land.player.LandPlayer
import net.onelitefeather.pandorascluster.land.position.HomePosition
import net.onelitefeather.pandorascluster.land.position.dummyHomePosition
import org.bukkit.Bukkit
import org.bukkit.Chunk
import org.bukkit.Material
import org.hibernate.Hibernate
import java.util.*

@Entity
data class Land(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @OneToOne
    val owner: LandPlayer? = null,

    @OneToOne
    val homePosition: HomePosition = dummyHomePosition(),

    @OneToMany(fetch = FetchType.EAGER, mappedBy = "land")
    val landMembers: List<LandMember> = arrayListOf(),

    @OneToMany(fetch = FetchType.EAGER, mappedBy = "land")
    val chunks: List<ChunkPlaceholder> = arrayListOf(),

    @OneToMany(fetch = FetchType.EAGER, mappedBy = "land")
    val flags: List<LandFlagEntity> = arrayListOf(),

    @Column
    val world: String = "world",

    @Column
    val x: Int = -1,

    @Column
    val z: Int = -1,
) {

    fun isOwner(uniqueId: UUID): Boolean = owner?.getUniqueId() == uniqueId

    fun getLandFlag(landFlag: LandFlag): LandFlagEntity =
        flags.find { it.name?.uppercase() == landFlag.name.uppercase() } ?: getDefaultFlag(landFlag).copy(land = this)


    fun getMergedChunk(chunkIndex: Long): ChunkPlaceholder? =
        chunks.firstOrNull { chunkPlaceholder -> chunkPlaceholder.chunkIndex == chunkIndex }

    fun getLandMember(uuid: UUID): LandMember? =
        landMembers.firstOrNull { landMember -> landMember.member?.getUniqueId() == uuid }

    fun hasMemberPermission(memberId: UUID, permission: Permission): Boolean {
        val bukkitPlayer = Bukkit.getPlayer(memberId) ?: return false
        return bukkitPlayer.hasPermission(permission)
    }

    fun hasMemberAccess(uuid: UUID): Boolean {
        if (isOwner(uuid)) return true
        val landOwner = owner ?: return false
        val landMember = getLandMember(uuid) ?: return false
        if (landMember.role == LandRole.MEMBER && !landOwner.isOnline()) return false
        return landMember.role.access
    }

    fun hasAccess(uuid: UUID): Boolean {
        if (isOwner(uuid)) return true
        if (hasMemberPermission(uuid, Permission.OWNED_CHUNK)) return true
        val landOwner = owner ?: return false
        val landMember = getLandMember(uuid) ?: return false
        if (landMember.role == LandRole.MEMBER && !landOwner.isOnline()) return false
        return landMember.role.access
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) return false
        other as Land

        return id != null && id == other.id
    }

    override fun hashCode(): Int = javaClass.hashCode()

    @Override
    override fun toString(): String {
        return this::class.simpleName +
                "(id = $id , owner = $owner , homePosition = $homePosition , " +
                "world = $world , x = $x , z = $z )"
    }

    fun isBanned(uniqueId: UUID): Boolean {
        val member = getLandMember(uniqueId) ?: return false
        return member.role == LandRole.BANNED
    }

    fun isAllowUse(material: Material): Boolean = getUseMaterials().contains(material)

    fun getUseMaterials(): List<Material> {
        if (!hasFlag(LandFlag.USE)) return emptyList()
        val value = getLandFlag(LandFlag.USE).getValue<String>() ?: return emptyList()

        if (!value.contains(",")) {
            val material = Material.matchMaterial(value)
            return if (material != null) listOf(material) else emptyList()
        } else {
            val args = value.split(",")
            val list = arrayListOf<Material>()
            for (materialName in args) {
                val material = Material.matchMaterial(materialName.uppercase()) ?: continue
                list.add(material)
            }
        }

        return emptyList()
    }

    fun isMerged() = chunks.map { chunkPlaceholder -> { chunkPlaceholder.chunkIndex != Bukkit.getWorld(world)?.getChunkAt(x, z)?.chunkKey } }.isNotEmpty()

    fun hasFlag(landFlag: LandFlag): Boolean = flags.any { landFlagEntity -> landFlagEntity.name == landFlag.name }

    fun isAdmin(playerId: UUID): Boolean {
        val member = getLandMember(playerId) ?: return false
        return member.role == LandRole.ADMIN
    }

    fun isMergedChunk(chunk: Chunk): Boolean = getMergedChunk(chunk.chunkKey) != null
}